central/delegatedregistryconfig/service/service.go

```diff
@@ -0,0 +1,227 @@
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	clusters, err := s.getClusters(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving clusters")
+	}
+
+	if len(clusters) == 0 {
+		return nil, status.Error(codes.NotFound, "no valid clusters found")
+	}
+
+	return &v1.DelegatedRegistryClustersResponse{
+		Clusters: clusters,
+	}, nil
+}
+
+// PutConfig updates Central's delegated registry config
+func (s *serviceImpl) PutConfig(ctx context.Context, config *storage.DelegatedRegistryConfig) (*storage.DelegatedRegistryConfig, error) {
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	if err := s.validate(ctx, config); err != nil {
+		return nil, errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	if err := s.dataStore.UpsertConfig(ctx, config); err != nil {
+		return nil, errors.Wrap(err, "upserting config")
+	}
+
+	return config, nil
+}
+
+func (s *serviceImpl) validate(ctx context.Context, config *storage.DelegatedRegistryConfig) error {
+	if config == nil {
+		// this block not reachable via GRPC-gateway invocations
+		return errors.New("config missing")
+	}
+
+	if config.EnabledFor == storage.DelegatedRegistryConfig_NONE {
+		// ignore rest of config
+		return nil
+	}
+
+	errorList := errorhelpers.NewErrorList("Validation")
+	// validate the default cluster id
+	if config.DefaultClusterId == "" {
+		errorList.AddStrings("defaultClusterId required if enabledFor is not NONE")
+		return errorList.ToError()
+	}
+
+	// get the valid clusters
+	validClusters, err := s.getValidClusterIds(ctx)
+	if err != nil {
+		return errors.Wrap(err, "unable to validate config")
+	}
+
+	if !validClusters.Contains(config.DefaultClusterId) {
+		errorList.AddStrings(fmt.Sprintf("default cluster %q is not a valid cluster", config.DefaultClusterId))
+	}
+
+	// validate the registries / clusters
+	for i, r := range config.Registries {
+
+		// if a cluster id was provided, check if its valid
+		if r.ClusterId != "" && !validClusters.Contains(r.ClusterId) {
+			errorList.AddStrings(fmt.Sprintf("cluster %q is not valid at index %v", r.ClusterId, i))
+		}
+
+		if r.RegistryPath == "" {
+			errorList.AddStrings(fmt.Sprintf("missing registry path at index %v", i))
+		}
+	}
+
+	return errorList.ToError()
+}
+
+// getClusters returns all clusters annotated with a flag indicating if cluster is valid
+// for use as a delegation target. All clusters are returned instead of just valid clusters
+// so that a consumer (ie: the UI) can show the friendly name of clusters that may no longer
+// be valid but once were
+func (s *serviceImpl) getClusters(ctx context.Context) ([]*v1.DelegatedRegistryCluster, error) {
+	clusters, err := s.clusterDataStore.GetClusters(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving clusters")
+	}
+
+	if len(clusters) == 0 {
+		return nil, nil
+	}
+
+	var res []*v1.DelegatedRegistryCluster
```

Comment:
since we know that the size of res will equal the size of clusters, you can presize this slice which makes appending faster because it won't resize the slice

res := make([]*v1.DelegatedRegistryCluster, 0, len(clusters))

Reply:
good catch, missed this one - originally started with not including 'invalid' clusters, so the size wasn't known ahead of time - will make the change

Reply:
fixed, initialized length instead of capacity, from what I understand thats slightly faster